directembedding/src/main/scala/ch/epfl/directembedding/transformers/EmbeddingTransformation.scala

```diff
@@ -0,0 +1,237 @@
+
+class reifyAs(to: Any) extends scala.annotation.StaticAnnotation
+
+/**
+ * Reify on the lifted "self" object with the same method name.
+ *
+ * Example:
+ *
+ *   direct.Query.take(1) => lift(direct.Query).take(lift(1))
+ *
+ * where the `take` in direct.Query has @reifyAsInvoked.
+ *
+ * An experimental reification that may be useful for existing
+ * deep DSLs, such as Slick. Instead of reifiying into a custom
+ * method (endpoint), the existing underlying API on the deep DSL
+ * is invoked directly.
+ */
+class reifyAsInvoked extends scala.annotation.StaticAnnotation
+
+/**
+ * Same as [[reifyAsInvoked]] except it can be annotated on modules.
+ *
+ * @param obj The object that the method is invoked on, "self".
+ */
+class reifyAsInvokedFrom(obj: Any) extends scala.annotation.StaticAnnotation
+
+/**
+ * No reificaation, the tree is passed through.
+ *
+ * This can be useful if you generate trees in the preprocessing
+ * step (e.g., type providers) that are already part of the deep DSL.
+ */
+class passThrough extends scala.annotation.StaticAnnotation
+
+trait EmbeddingTransformation extends DirectEmbeddingModule
+  with DirectEmbeddingUtils
+  with ReificationTransformation
+  with TypeHelper {
+  import c.universe._
+
+  object EmbeddingTransformer extends (Tree => Tree) {
+    def apply(tree: Tree) = {
+      logStarred("Before ReifyAsEmbedding")
+      new EmbeddingTransformer().apply(tree)
+    }
+  }
+
+  final class EmbeddingTransformer extends Transformer {
+    def apply(tree: Tree): Tree = transform(tree)
+
+    private def reifyAnnotation(methodSym: Symbol): Option[Annotation] =
+      methodSym.annotations.find(_.tree.tpe <:< c.typeOf[reifyAs])
+
+    private def annotations(sym: Symbol, tree: Tree): List[Annotation] = {
+      fallbackAnnotation(sym, tree).toList ::: sym.annotations
+    }
+
+    private def reify(lhs: Tree, tree: Tree, extraTargs: List[Tree], extraArgs: List[Tree]): Tree = {
+      val sym = symbol(lhs)
+      val reification: List[Reification] = annotations(sym, tree).collect {
+        case a if a.tree.tpe =:= c.typeOf[reifyAs] =>
+          ReifyAsReification(
+            annotationBody(a),
+            typeArgs(lhs) ::: extraTargs,
+            args(lhs, extraArgs))
+        case a if a.tree.tpe =:= c.typeOf[reifyAsInvoked] =>
+          val s = self(lhs).getOrElse {
+            c.abort(tree.pos, "reifyAsInvoked cannot be annotated to an object method")
+          }
+          ReifyAsInvokedReification(transform(s), sym.name.toTermName, extraArgs.map(transform))
+        case a if a.tree.tpe <:< c.typeOf[reifyAsInvokedFrom] =>
+          val s = a.tree.children.tail.headOption.getOrElse {
+            c.abort(tree.pos, "Missing argument for @reifyAsInvokedFrom annotation")
+          }
+          ReifyAsInvokedReification(s, sym.name.toTermName, extraArgs.map(transform))
+        case a if a.tree.tpe <:< c.typeOf[passThrough] =>
```

Comment:
This was the easiest solution I could think of to embed the `driver.Table` type provider for slick-direct. I am open for suggestions on alternative approaches.